{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# scVI\n",
    "\n",
    "**scVI** [^ref1] (single-cell Variational Inference; Python class {class}`~scvi.model.SCVI`) posits a flexible generative model of scRNA-seq count data that can subsequently\n",
    "be used for many common downstream tasks.\n",
    "\n",
    "The advantages of scVI are:\n",
    "\n",
    "-   Comprehensive in capabilities.\n",
    "-   Scalable to very large datasets (>1 million cells).\n",
    "\n",
    "The limitations of scVI include:\n",
    "\n",
    "-   Effectively requires a GPU for fast inference.\n",
    "-   Latent space is not interpretable, unlike that of a linear method.\n",
    "\n",
    "```{topic} Tutorials:\n",
    "\n",
    "-   {doc}`/tutorials/notebooks/quick_start/api_overview`\n",
    "-   {doc}`/tutorials/notebooks/scrna/harmonization`\n",
    "-   {doc}`/tutorials/notebooks/scrna/scarches_scvi_tools`\n",
    "-   {doc}`/tutorials/notebooks/scrna/scvi_in_R`\n",
    "```\n",
    "\n",
    "## Preliminaries\n",
    "\n",
    "scVI takes as input a scRNA-seq gene expression matrix $X$ with $N$ cells and $G$ genes.\n",
    "Additionally, a design matrix $S$ containing $p$ observed covariates, such as day, donor, etc, is an optional input.\n",
    "While $S$ can include both categorical covariates and continuous covariates, in the following, we assume it contains only one\n",
    "categorical covariate with $K$ categories, which represents the common case of having multiple batches of data.\n",
    "\n",
    "## Generative process\n",
    "\n",
    "scVI posits that the observed UMI counts for cell $n$ and gene $g$, $x_{ng}$, are generated\n",
    "by the following process:\n",
    "\n",
    "```{math}\n",
    ":nowrap: true\n",
    "\n",
    "\\begin{align}\n",
    " z_n &\\sim {\\mathrm{Normal}}\\left( {0,I} \\right) \\\\\n",
    " \\ell_n &\\sim \\mathrm{LogNormal}\\left( \\ell_\\mu^\\top s_n ,\\ell_{\\sigma^2}^\\top s_n \\right) \\\\\n",
    " \\rho _n &= f_w\\left( z_n, s_n \\right) \\\\\n",
    " \\pi_{ng} &= f_h^g(z_n, s_n) \\\\\n",
    " x_{ng} &\\sim \\mathrm{ObservationModel}(\\ell_n \\rho_n, \\theta_g, \\pi_{ng})\n",
    " \\end{align}\n",
    "```\n",
    "\n",
    "Succintly, the gene expression for each gene depends on a latent variable $z_n$ that is cell-specific.\n",
    "The prior parameters $\\ell_\\mu$ and $\\ell_{\\sigma^2}$ are computed per batch as the mean and variance of the log library size over cells.\n",
    "The expression data are generated from a count-based likelihood distribution, which here, we denote as the $\\mathrm{ObservationModel}$.\n",
    "While by default the $\\mathrm{ObservationModel}$ is a $\\mathrm{ZeroInflatedNegativeBinomial}$ (ZINB) distribution parameterized by its mean, inverse dispersion, and non-zero-inflation probability, respectively,\n",
    "users can pass `gene_likelihood = \"negative_binomial\"` to {class}`~scvi.model.SCVI`, for example, to use a simpler $\\mathrm{NegativeBinomial}$ distribution.\n",
    "\n",
    "The generative process of scVI uses two neural networks:\n",
    "\n",
    "```{math}\n",
    ":nowrap: true\n",
    "\n",
    "\\begin{align}\n",
    "   f_w(z_n, s_n) &: \\mathbb{R}^{d} \\times \\{0, 1\\}^K \\to \\Delta^{G-1}\\\\\n",
    "   f_h(z_n, s_n) &: \\mathbb{R}^d \\times \\{0, 1\\}^K \\to (0, 1)^T\n",
    "\\end{align}\n",
    "```\n",
    "\n",
    "which respectively decode the denoised gene expression and non-zero-inflation probability (only if using ZINB).\n",
    "\n",
    "This generative process is also summarized in the following graphical model:\n",
    "\n",
    ":::{figure} figures/scvi_annotated_graphical_model.png\n",
    ":align: center\n",
    ":alt: scVI graphical model\n",
    ":class: img-fluid\n",
    "\n",
    "scVI graphical model for the ZINB likelihood model. Note that this graphical model contains more latent variables than the presentation above. Marginalization of these latent variables leads to the ZINB observation model (math shown in publication supplement).\n",
    ":::\n",
    "\n",
    "The latent variables, along with their description are summarized in the following table:\n",
    "\n",
    "```{eval-rst}\n",
    ".. list-table::\n",
    "   :widths: 20 90 15\n",
    "   :header-rows: 1\n",
    "\n",
    "   * - Latent variable\n",
    "     - Description\n",
    "     - Code variable (if different)\n",
    "   * - :math:`z_n \\in \\mathbb{R}^d`\n",
    "     - Low-dimensional representation capturing the state of a cell.\n",
    "     - N/A\n",
    "   * - :math:`\\rho_n \\in \\Delta^{G-1}`\n",
    "     - Denoised/normalized gene expression. This is a vector that sums to 1 within a cell, unless `size_factor_key is not None` in :class:`~scvi.model.SCVI.setup_anndata`, in which case this is only forced to be non-negative via softplus.\n",
    "     - ``px_scale``\n",
    "   * - :math:`\\ell_n \\in (0, \\infty)`\n",
    "     - Library size for RNA. Here it is modeled as a latent variable, but the recent default for scVI is to treat library size as observed, equal to the total RNA UMI count of a cell. This can be controlled by passing ``use_observed_lib_size=False`` to :class:`~scvi.model.SCVI`. The library size can also be set manually using `size_factor_key` in :class:`~scvi.model.SCVI.setup_anndata`.\n",
    "     - N/A\n",
    "   * - :math:`\\theta_g \\in (0, \\infty)`\n",
    "     - Inverse dispersion for negative binomial. This can be set to be gene/batch specific for example (and would thus be :math:`\\theta_{kg}`), by passing ``dispersion=\"gene-batch\"`` during model intialization. Note that ``px_r`` also refers to the underlying real-valued torch parameter that is then exponentiated on every forward pass of the model.\n",
    "     - ``px_r``\n",
    "```\n",
    "\n",
    "## Inference\n",
    "\n",
    "scVI uses variational inference and specifically auto-encoding variational bayes (see {doc}`/user_guide/background/variational_inference`) to learn both the model parameters (the\n",
    "neural network params, dispersion params, etc.) and an approximate posterior distribution with the following factorization:\n",
    "\n",
    "```{math}\n",
    ":nowrap: true\n",
    "\n",
    "\\begin{align}\n",
    "   q_\\eta(z_n, \\ell_n \\mid x_n) :=\n",
    "   q_\\eta(z_n \\mid x_n, s_n)q_\\eta(\\ell_n \\mid x_n).\n",
    "\\end{align}\n",
    "```\n",
    "\n",
    "Here $\\eta$ is a set of parameters corresponding to inference neural networks (encoders), which we do not describe in detail here,\n",
    "but are described in the scVI paper. The underlying class used as the encoder for scVI is {class}`~scvi.nn.Encoder`.\n",
    "In the case of `use_observed_lib_size=True`, $q_\\eta(\\ell_n \\mid x_n)$ can be written as a point mass on the observed library size.\n",
    "\n",
    "It it important to note that by default, scVI only\n",
    "receives the expression data as input (i.e., not the observed cell-level covariates).\n",
    "Empirically, we have not seen much of a difference by having the encoder take as input the concatenation of these items (i.e., $q_\\eta(z_n, \\ell_n \\mid x_n, s_n)$, but users can control it manually by passing\n",
    "`encode_covariates=True` to {class}`scvi.model.SCVI`.\n",
    "\n",
    "## Tasks\n",
    "\n",
    "Here we provide an overview of some of the tasks that scVI can perform. Please see {class}`scvi.model.SCVI` for the full API reference.\n",
    "\n",
    "### Dimensionality reduction\n",
    "\n",
    "For dimensionality reduction, the mean of the approximate posterior $q_\\eta(z_n \\mid x_n, s_n)$ is returned by default.\n",
    "This is achieved using the method:\n",
    "\n",
    "```\n",
    ">>> latent = model.get_latent_representation()\n",
    ">>> adata.obsm[\"X_scvi\"] = latent\n",
    "```\n",
    "\n",
    "Users may also return samples from this distribution, as opposed to the mean by passing the argument `give_mean=False`.\n",
    "The latent representation can be used to create a nearest neighbor graph with scanpy with:\n",
    "\n",
    "```\n",
    ">>> import scanpy as sc\n",
    ">>> sc.pp.neighbors(adata, use_rep=\"X_scvi\")\n",
    ">>> adata.obsp[\"distances\"]\n",
    "```\n",
    "\n",
    "### Transfer learning\n",
    "\n",
    "A scVI model can be pre-trained on reference data and updated with query data using {func}`~scvi.model.SCVI.load_query_data`, which then facilitates transfer of metadata like cell type annotations. See the {doc}`/user_guide/background/transfer_learning` guide for more information.\n",
    "\n",
    "### Normalization/denoising/imputation of expression\n",
    "\n",
    "In {func}`~scvi.model.SCVI.get_normalized_expression` scVI returns the expected value of $\\rho_n$ under the approximate posterior. For one cell $n$, this can be written as:\n",
    "\n",
    "```{math}\n",
    ":nowrap: true\n",
    "\n",
    "\\begin{align}\n",
    "   \\mathbb{E}_{q_\\eta(z_n \\mid x_n)}\\left[\\ell_n'f_w\\left( z_n, s_n \\right) \\right],\n",
    "\\end{align}\n",
    "```\n",
    "\n",
    "where $\\ell_n'$ is by default set to 1. See the `library_size` parameter for more details. The expectation is approximated using Monte Carlo, and the number of samples can be passed as an argument in the code:\n",
    "\n",
    "```\n",
    ">>> model.get_normalized_expression(n_samples=10)\n",
    "```\n",
    "\n",
    "By default the mean over these samples is returned, but users may pass `return_mean=False` to retrieve all the samples.\n",
    "\n",
    "Notably, this function also has the `transform_batch` parameter that allows counterfactual prediction of expression in an unobserved batch. See the {doc}`/user_guide/background/counterfactual_prediction` guide.\n",
    "\n",
    "### Differential expression\n",
    "\n",
    "Differential expression analysis is achieved with {func}`~scvi.model.SCVI.differential_expression`. scVI tests differences in magnitude of $f_w\\left( z_n, s_n \\right)$. More info is in {doc}`/user_guide/background/differential_expression`.\n",
    "\n",
    "### Data simulation\n",
    "\n",
    "Data can be generated from the model using the posterior predictive distribution in {func}`~scvi.model.SCVI.posterior_predictive_sample`.\n",
    "This is equivalent to feeding a cell through the model, sampling from the posterior\n",
    "distributions of the latent variables, retrieving the likelihood parameters (of $p(x \\mid z, s)$), and finally, sampling from this distribution.\n",
    "\n",
    "[^ref1]:\n",
    "    Romain Lopez, Jeffrey Regier, Michael Cole, Michael I. Jordan, Nir Yosef (2018),\n",
    "    _Deep generative modeling for single-cell transcriptomics_,\n",
    "    [Nature Methods](https://www.nature.com/articles/s41592-018-0229-2.epdf?author_access_token=5sMbnZl1iBFitATlpKkddtRgN0jAjWel9jnR3ZoTv0P1-tTjoP-mBfrGiMqpQx63aBtxToJssRfpqQ482otMbBw2GIGGeinWV4cULBLPg4L4DpCg92dEtoMaB1crCRDG7DgtNrM_1j17VfvHfoy1cQ%3D%3D).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
